File: code/platform/src/main/java/org/fbme/ide/platform/editor/ProjectionChooserComboBoxAction.java
package org.fbme.ide.platform.editor;

import com.intellij.ide.DataManager;
import com.intellij.openapi.actionSystem.*;
import com.intellij.openapi.actionSystem.ex.ComboBoxAction;
import com.intellij.openapi.vfs.VirtualFile;
import jetbrains.mps.nodefs.MPSNodeVirtualFile;
import jetbrains.mps.project.Project;
import jetbrains.mps.workbench.MPSDataKeys;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.mps.openapi.model.SNode;

import javax.swing.*;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class ProjectionChooserComboBoxAction extends ComboBoxAction {

    private final HeaderedNodeEditor myEditor;

    public ProjectionChooserComboBoxAction(HeaderedNodeEditor editor) {
        myEditor = editor;
    }

    @Override
    public void update(@NotNull AnActionEvent e) {
        Presentation presentation = e.getPresentation();
        presentation.setText(myEditor.getCurrentProjection().getName());
    }

    @NotNull
    @Override
    protected DefaultActionGroup createPopupActionGroup(JComponent component) {
        DataContext context = DataManager.getInstance().getDataContext(component);
        final DefaultActionGroup group = new DefaultActionGroup();
        final VirtualFile vf = context.getData(LangDataKeys.VIRTUAL_FILE);
        final Project project = context.getData(MPSDataKeys.MPS_PROJECT);

        if (project != null && vf instanceof MPSNodeVirtualFile) {
            project.getModelAccess().runReadAction(() -> {
                final SNode node = Objects.requireNonNull(((MPSNodeVirtualFile) vf).getNode());

                List<EditorProjectionController> controllers = EditorProjectionControllerRegistry.getInstance()
                        .getFactories().stream()
                        .filter(it -> it.isApplicable(node))
                        .map(it -> it.create(node, project))
                        .collect(Collectors.toList());

                for (EditorProjectionController controller : controllers) {
                    group.addAll(controller.getChooseProjectionActions());
                }
                group.addSeparator();
                for (EditorProjectionController controller : controllers) {
                    group.addAll(controller.getCreateProjectionActions());
                }

            });
        }
        return group;
    }

}

File: code/platform/src/main/kotlin/org/fbme/ide/platform/editor/ProjectionChooserComboBoxAction.kt
package org.fbme.ide.platform.editor

import com.intellij.ide.DataManager
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.DefaultActionGroup
import com.intellij.openapi.actionSystem.LangDataKeys
import com.intellij.openapi.actionSystem.ex.ComboBoxAction
import jetbrains.mps.nodefs.MPSNodeVirtualFile
import jetbrains.mps.project.Project
import jetbrains.mps.workbench.MPSDataKeys
import javax.swing.JComponent

class ProjectionChooserComboBoxAction(private val myEditor: HeaderedNodeEditor) : ComboBoxAction() {
    override fun update(e: AnActionEvent) {
        val presentation = e.presentation
        presentation.text = myEditor.currentProjection!!.name
    }

    override fun createPopupActionGroup(component: JComponent): DefaultActionGroup {
        val context = DataManager.getInstance().getDataContext(component)
        val group = DefaultActionGroup()
        val vf = context.getData(LangDataKeys.VIRTUAL_FILE)
        val project: Project? = context.getData(MPSDataKeys.MPS_PROJECT)
        if (project != null && vf is MPSNodeVirtualFile) {
            project.modelAccess.runReadAction {
                val node = requireNotNull(vf.node)
                val controllers: List<EditorProjectionController> =
                    EditorProjectionControllerRegistry.instance.factories
                        .filter { it.isApplicable(node) }
                        .map { it.create(node, project) }
                for (controller in controllers) {
                    group.addAll(controller.chooseProjectionActions)
                }
                group.addSeparator()
                for (controller in controllers) {
                    group.addAll(controller.createProjectionActions)
                }
            }
        }
        return group
    }
}
